### PR TITLE
CCv0: Optimize integrity device creation by avoiding a full device clear

### LINKED ISSUE #5329: CCv0: Optimize ephemeral storage creation
**Which feature do you think can be improved?**

Currently, ephemeral storage creation is slow. The `init_trusted_storage.sh` script takes about 90 seconds to complete for a 16GiB disk.
In Azure, we'd like to provide an experience close to current AKS containers - where the rw layer starts off with 128GiB at the lower-end and goes all the way up to 2TiB.  Therefore, it would be beneficial to optimize integrity device creation.

**How can it be improved?**

dm-integrity by default performs a full wipe of the device, and this is what takes up majority of the time. This is a known issue with dm-integrity captured in https://gitlab.com/cryptsetup/cryptsetup/-/issues/525.

That issue describes the problem and how to solve it.

**Additional Information**

**Before raising this enhancement request**

Have you looked at the [limitations document](https://github.com/kata-containers/kata-containers/blob/main/docs/Limitations.md)?
Yes.

### PR DESCRIPTION
**Pull request by anakrish**
Based on https://gitlab.com/cryptsetup/cryptsetup/-/issues/525

With this PR, 1 TiB ephemeral disk can be setup in ~9 seconds.

1. When --no-wipe is used, the device will have invalid checksums
2. mkfs.ext4 would fail on an un-wiped device due to reads of pages with invalid checksums
3. To make mkfs.ext4 work
   - Perform a dry run to figure out which sectors (pages) mkfs.ext4 will write to.
   - Perform direct writes to these pages to ensure that they will have valid checksums
   - Invoke mkfs.ext4 again to perform initialization 
4. Use lazy_journal_init option with mkfs.ext4 to lazily initialize the journal.
  According to the man pages,
  "This speeds up file system initialization  noticeably, but carries some small
   risk if the system crashes before the  journal  has  been overwritten  entirely
   one time."
  Since the storage is ephemeral, not expected to survive a system crash/power cycle,
  it is safe to use lazy_journal_init.

Fixes #5329 

Signed-off-by: Anand Krishnamoorthi <anakrish@microsoft.com>

### CODE REVIEW

**Comment by wainersm** [comment]
Hi @anakrish , first of all, thanks for your contribution & welcome to the kata containers community.

Unfortunately this PR is out of my expertise area but I will try to well with what I can.

If this pull request introduce a feature specific for Confidential Containers, which seems the case, then the `CCv0` is the right branch. Otherwise, you better send to the `main` branch and it will be picked up on the weekly rebases we do.

Let's include @arronwy on the loop as we implemented the initial ephemeral storage support.

Lastly, I will run the static checks to ensure the path is OK.

**Comment by wainersm** [comment]
It should be created a enhancement request as an issue on this repository then reference the issue number in your commit message (`Fixes #xxxx`)

**Comment by anakrish** [comment]
@wainersm Do I need to add a `wip` in title? 
I got notified of the following failure message:
> [WIP Check](https://github.com/kata-containers/kata-containers/actions/runs/3170502319/jobs/5207715467#step:2:6)
Blocking keywords in title: wip

**Comment by wainersm** [comment]
> @wainersm Do I need to add a `wip` in title? I got notified of the following failure message:
> 

The other way around. Apparently there is a bug on the action that implements the WIP checker. On the commit subject, it interprets the `wip` on  `--no-wipe` as an indication of WiP (work in progress) so the checker fail (any commit with `wip` in subject will block the PR merge).

@anakrish could you please re-phase the commit and PR's subject to remove the `--no-wipe` string? I think it will make the checker happy. In meanwhile I will open a bug. 
 
> > [WIP Check](https://github.com/kata-containers/kata-containers/actions/runs/3170502319/jobs/5207715467#step:2:6)
> > Blocking keywords in title: wip

**Comment by wainersm** [comment]
> @wainersm Do I need to add a `wip` in title? I got notified of the following failure message:
> 
> > [WIP Check](https://github.com/kata-containers/kata-containers/actions/runs/3170502319/jobs/5207715467#step:2:6)
> > Blocking keywords in title: wip
> 
> Maybe `--no-wipe` is getting interpreted as `wip` ?

yes, exactly :(

**Comment by anakrish** [comment]
Thanks for the quick feedback.
In the workflow checks I see that the following failed:
> jenkins-ci-tdx-x86_64-cc_cri_containerd_k8s-cloud-hypervisor
> jenkins-ci-tdx-x86_64-cc_cri_containerd_k8s-qemu

When I click the Details button, it takes me to a non-accessible IP address.

I'd appreciate it if you could let me know what the failures are.

**Comment by wainersm** [comment]
> Thanks for the quick feedback. In the workflow checks I see that the following failed:
> 
> > jenkins-ci-tdx-x86_64-cc_cri_containerd_k8s-cloud-hypervisor
> > jenkins-ci-tdx-x86_64-cc_cri_containerd_k8s-qemu
> 
> When I click the Details button, it takes me to a non-accessible IP address.
>

The TDX jobs are maintained by Intel and run behind a firewall, that's why we don't have access to the logs.
 
> I'd appreciate it if you could let me know what the failures are.

I believe those jobs failed due to an infra error as explained in https://github.com/kata-containers/kata-containers/issues/5326 . I will wait all the static jobs to finish then I will trigger the Jenkins jobs. Before merging we can try to re-run the TDX jobs if they are back running.

**Comment by anakrish** [comment]
@wainersm Do let me know if something is needed from my end.

Is ephemeral storage enabled in CCv0? @mikbras tried to pull a large image (python:latest) and ran into issue #5408.
When investigating, he observed that there was no /dev/mapper within the kata uvm and agent.data_integrity didn't seem to have any effect.

**Review by arronwy — APPROVED** [review]
Thanks @anakrish , I verified this PR, now with integrity enabled, the boot time is close to the config without integrity enable, thanks a lot!.

**Comment by wainersm** [comment]
> Thanks @anakrish , I verified this PR, now with integrity enabled, the boot time is close to the config without integrity enable, thanks a lot!.

Thanks @arronwy ! How did you test it? Can we try to make an automated test to run on CI?

**Review by wainersm — APPROVED** [review]

**Comment by arronwy** [comment]
> > Thanks @anakrish , I verified this PR, now with integrity enabled, the boot time is close to the config without integrity enable, thanks a lot!.
> 
> Thanks @arronwy ! How did you test it? Can we try to make an automated test to run on CI?

Hi @wainersm , I tested it with `open-local` CSI driver, we may need an separate disk and install a CSI driver which support local  disk, it may a bit hard to setup in current CI.

**Comment by anakrish** [comment]
> > > Thanks @anakrish , I verified this PR, now with integrity enabled, the boot time is close to the config without integrity enable, thanks a lot!.
> > 
> > 
> > Thanks @arronwy ! How did you test it? Can we try to make an automated test to run on CI?
> 
> Hi @wainersm , I tested it with `open-local` CSI driver, we may need an separate disk and install a CSI driver which support local disk, it may a bit hard to setup in current CI.

Would it help if we tweak the kata-cc shim to:
- Create a sparse disk on host 
- Hot plug the disk to the kata-uvm?

That way each kata-uvm will have a disk that it could use if needed for ephemeral storage.

**Pull request merged**
